x/capability/keeper/keeper.go

```diff
@@ -0,0 +1,244 @@
+	// tracking, and authenticating capabilities at runtime. During application
+	// initialization, the keeper can be hooked up to modules through unique function
+	// references so that it can identify the calling module when later invoked.
+	//
+	// When the initial state is loaded from disk, the keeper allows the ability to
+	// create new capability keys for all previously allocated capability identifiers
+	// (allocated during execution of past transactions and assigned to particular modes),
+	// and keep them in a memory-only store while the chain is running.
+	//
+	// The keeper allows the ability to create scoped sub-keepers which are tied to
+	// a single specific module.
+	Keeper struct {
+		cdc           *codec.Codec
+		storeKey      sdk.StoreKey
+		memKey        sdk.StoreKey
+		scopedModules map[string]struct{}
+		sealed        bool
+	}
+
+	// ScopedKeeper defines a scoped sub-keeper which is tied to a single specific
+	// module provisioned by the capability keeper. Scoped keepers must be created
+	// at application initialization and passed to modules, which can then use them
+	// to claim capabilities they receive and retrieve capabilities which they own
+	// by name, in addition to creating new capabilities & authenticating capabilities
+	// passed by other modules.
+	ScopedKeeper struct {
+		cdc      *codec.Codec
+		storeKey sdk.StoreKey
+		memKey   sdk.StoreKey
+		module   string
+	}
+)
+
+func NewKeeper(cdc *codec.Codec, storeKey, memKey sdk.StoreKey) Keeper {
+	return Keeper{
+		cdc:           cdc,
+		storeKey:      storeKey,
+		memKey:        memKey,
+		scopedModules: make(map[string]struct{}),
+		sealed:        false,
+	}
+}
+
+// ScopeToModule attempts to create and return a ScopedKeeper for a given module
+// by name. It will panic if the keeper is already sealed or if the module name
+// already has a ScopedKeeper.
+func (k Keeper) ScopeToModule(moduleName string) ScopedKeeper {
+	if k.sealed {
+		panic("cannot scope to module via a sealed capability keeper")
+	}
+
+	if _, ok := k.scopedModules[moduleName]; ok {
+		panic(fmt.Sprintf("cannot create multiple scoped keepers for the same module name: %s", moduleName))
+	}
+
+	k.scopedModules[moduleName] = struct{}{}
+
+	return ScopedKeeper{
+		cdc:      k.cdc,
+		storeKey: k.storeKey,
+		memKey:   k.memKey,
+		module:   moduleName,
+	}
+}
+
+// InitializeAndSeal loads all capabilities from the persistent KVStore into the
+// in-memory store and seals the keeper to prevent further modules from creating
+// a scoped keeper. InitializeAndSeal must be called once after the application
+// state is loaded.
+func (k Keeper) InitializeAndSeal(ctx sdk.Context) {
+	if k.sealed {
+		panic("cannot initialize and seal an already sealed capability keeper")
+	}
+
+	memStore := ctx.KVStore(k.memKey)
+	memStoreType := memStore.GetStoreType()
+
+	if memStoreType != sdk.StoreTypeMemory {
+		panic(fmt.Sprintf("invalid memory store type; got %d, expected: %d", memStoreType, sdk.StoreTypeMemory))
```

Comment:
nit. Imo it'd be useful to have a `String()` method to create a better error message here

Reply:
already done. I haven't pushed all the code.